File: 프로그래머스/lv3/지준혁/다단계 칫솔 판매.cpp
#include <string>
#include <vector>
#include <unordered_map>
#include <iostream>

using namespace std;

void recur(string name, int income, unordered_map<string, int>& profits, unordered_map<string, string>& parent) {
    if (name == "-" || income < 1)
        return;
    int fee = income * 0.1;
    profits[name] += income - fee;
    recur(parent[name], fee, profits, parent);
}

vector<int> solution(vector<string> enroll, vector<string> referral, vector<string> seller, vector<int> amount) {
    
    unordered_map<string, string> parent;
    unordered_map<string, int> profits;
    
    for (auto& e : enroll) {
        profits[e] = 0;
    }
    for (int i = 0; i < enroll.size(); ++i) {
        parent[enroll[i]] = referral[i];
    }

    for (int i = 0; i < seller.size(); ++i) {
        int total_profit = 100 * amount[i];
        recur(seller[i], 100 * amount[i], profits, parent);
    }
    
    vector<int> ans;
    for (auto& e : enroll) {
        ans.push_back(profits[e]);
    }
    return ans;
}

File: 프로그래머스/lv3/최원준/다단계 칫솔 판매.py
'''
1. 아이디어 :
    dfs를 사용해서 풀 수 있다
2. 시간복잡도 :
    O( n^2 )
3. 자료구조 :
    해시맵
'''

from collections import defaultdict
import math

def solution(enroll, referral, seller, amount):
    def dfs(name, income):
        if name == "-" or income == 0:
            return
        commission = math.floor(income * 0.1)
        profit[name] += income - commission
        dfs(par[name], commission)

    par = {enroll[i] : referral[i] for i in range(len(enroll))}
    profit = defaultdict(int)

    for i in range(len(seller)):
        dfs(seller[i], amount[i] * 100)

    return [profit[person] for person in enroll]